File: content/examples/other/leetcode/problems/decode-ways/example-1/solution.py
def num_decodings_top_down(s):
    memo = {}

    def dfs(i):
        if i == len(s):
            return 1
        if s[i] == "0":
            return 0
        if i in memo:
            return memo[i]

        ways = dfs(i + 1)
        if i + 1 < len(s):
            two = int(s[i:i + 2])
            if 10 <= two <= 26:
                ways += dfs(i + 2)
        memo[i] = ways
        return ways

    return dfs(0)


if __name__ == "__main__":
    print(num_decodings_top_down("226"))

File: content/examples/other/leetcode/problems/decode-ways/example-1/solution.js
function numDecodingsTopDown(s) {
  const memo = new Map();
  function dfs(i) {
    if (i === s.length) return 1;
    if (s[i] === "0") return 0;
    if (memo.has(i)) return memo.get(i);

    let ways = dfs(i + 1);
    if (i + 1 < s.length) {
      const two = Number(s.slice(i, i + 2));
      if (two >= 10 && two <= 26) ways += dfs(i + 2);
    }
    memo.set(i, ways);
    return ways;
  }
  return dfs(0);
}

if (require.main === module) {
  console.log(numDecodingsTopDown("226"));
}

module.exports = { numDecodingsTopDown };
